[quant][docs] Add fx graph mode quantization to quantization docs

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49211 [quant][docs] Add fx graph mode quantization to quantization docs**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25507480](https://our.internmc.facebook.com/intern/diff/D25507480)

jerryzh168 committed:
[quant][docs] Add fx graph mode quantization to quantization docs

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on docs/source/quantization.rst:
```diff
@@ -80,7 +80,10 @@ The corresponding implementation is chosen automatically based on the PyTorch bu
 ---------------------------------------
 
-There are three types of quantization supported in PyTorch:
+Eager Mode Quantization
```
Should we add a quick section for context on Eager vs FX Graph (why both, what are the tradeoffs, etc)?  Doesn't have to be this PR.

vkuzo commented on docs/source/quantization.rst:
```diff
@@ -332,6 +335,95 @@ To learn more about quantization aware training, please see the `QAT
 tutorial
 <https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html>`_.
 
+FX Graph Mode Quantization
```
nit: would it make sense to move this to a separate page, and link to it?  The Eager quantization page is already very long.

jerryzh168 replied:
should we create a folder and then split? I think we already have many quantization files under source/

vkuzo commented on docs/source/quantization.rst:
```diff
@@ -332,6 +335,95 @@ To learn more about quantization aware training, please see the `QAT
+
+The supported quantization types in FX Graph Mode Quantization are:
+- Post Training Quantization
+  - Weight Only Quantization
```
optional: we could skip this from the docs until we approach beta.

jerryzh168 replied:
would like to keep this since our apis assumes this classification.

vkuzo commented on docs/source/quantization.rst:
```diff
@@ -332,6 +335,95 @@ To learn more about quantization aware training, please see the `QAT
+- `prepare_fx`: prepares a model for post training quantization
+- `prepare_qat_fx`: prepares a model for quantization aware training
+- `convert_fx`: converts a prepared (and calibrated/trained) model to a quantized model
+- `fuse_fx`: This is not a function that automatically fuses module calls like Conv - BatchNorm into Conv. Currently only supports eval model.
```
optional: we could skip this it on this page, since it is inlined in the other APIs.

vkuzo commented on docs/source/quantization.rst:
```diff
@@ -332,6 +335,95 @@ To learn more about quantization aware training, please see the `QAT
+- `convert_fx`: converts a prepared (and calibrated/trained) model to a quantized model
+- `fuse_fx`: This is not a function that automatically fuses module calls like Conv - BatchNorm into Conv. Currently only supports eval model.
+
+There are multiple quantization types in post training quantization (weight only, dynamic and static) and the configuration is done through `qconfig_dict` (an argument of the `prepare_fx` function), and mixed quantization type is also supported with the api (e.g. quantize the conv modules statically and quantize the linear module dynamically).
```
optional: we could skip edge casey types until we have full docs (i.e we don't have to mention mixed quantization, etc)

vkuzo commented on docs/source/quantization.rst:
```diff
@@ -332,6 +335,95 @@ To learn more about quantization aware training, please see the `QAT
 <https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html>`_.
 
+FX Graph Mode Quantization
+^^^^^^^^^^^^^^^^^^^^^^^^^^
```
another nit: should we include "prototype" in the title

vkuzo reviewed (approved):
awesome!

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25507480](https://our.internmc.facebook.com/intern/diff/D25507480)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25507480](https://our.internmc.facebook.com/intern/diff/D25507480)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25507480](https://our.internmc.facebook.com/intern/diff/D25507480)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25507480](https://our.internmc.facebook.com/intern/diff/D25507480)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25507480](https://our.internmc.facebook.com/intern/diff/D25507480)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25507480](https://our.internmc.facebook.com/intern/diff/D25507480)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25507480](https://our.internmc.facebook.com/intern/diff/D25507480)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add fx graph mode quantization to quantization docs"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25507480](https://our.internmc.facebook.com/intern/diff/D25507480)

[ghstack-poisoned]

mruberry commented:
Unlanded. This broke several builds with mypy failures. The failures appear in the Github CI for this PR, too.

jerryzh168 commented:
I see, I thought it was fixed in master by https://github.com/pytorch/pytorch/blob/master/tools/codegen/gen.py#L471

jerryzh168 commented:
oh accidently removed the return in this PR
